test/tools/dshell.d

```diff
@@ -0,0 +1,290 @@
+public import std.meta;
+public import std.exception;
+public import std.array;
+public import std.string;
+public import std.format;
+public import std.path;
+public import std.file;
+public import std.stdio;
+public import std.process;
+
+struct Vars
+{
+    private static __gshared string[string] map;
+    static void set(string name, string value)
+    in { assert(value !is null); } do
+    {
+        const expanded = expand(value);
+        assert(expanded !is null, "codebug");
+        map[name] = expanded;
+    }
+    static string get(string name)
+    {
+        auto result = map.get(name, null);
+        if (result is null)
+            assert(0, "Unknown variable '" ~ name ~ "'");
+        return result;
+    }
+    static string opDispatch(string name)() { return get(name); }
+}
+
+alias requiredEnvVars = AliasSeq!(
+    "MODEL", "RESULTS_DIR",
+    "EXE", "OBJ",
+    "DMD", "DFLAGS",
+    "OS", "SEP", "DSEP",
+);
+alias allVars = AliasSeq!(
+    requiredEnvVars,
+    "TEST_DIR", "TEST_NAME",
+    "RESULTS_TEST_DIR",
+    "OUTPUT_BASE", "EXTRA_FILES",
+    "LIBEXT",
+);
+
+static foreach (var; allVars)
+{
+    mixin(`string ` ~ var ~ `() { return Vars.` ~ var ~ `; }`);
+}
+
+static this()
+{
+    static import dshellglue;
+
+    static foreach (var; requiredEnvVars)
+    {
+        mixin(`Vars.set("` ~ var ~ `", requireEnv("` ~ var ~ `"));`);
+    }
+
+    Vars.set("TEST_DIR", dshellglue.TEST_DIR);
+    Vars.set("TEST_NAME", dshellglue.TEST_NAME);
+    // reference to the resulting test_dir folder, e.g .test_results/runnable
+    Vars.set("RESULTS_TEST_DIR", buildPath(RESULTS_DIR, TEST_DIR));
+    // reference to the resulting files without a suffix, e.g. test_results/runnable/test123import test);
+    Vars.set("OUTPUT_BASE", buildPath(RESULTS_TEST_DIR, TEST_NAME));
+    // reference to the extra files directory
+    Vars.set("EXTRA_FILES", buildPath(TEST_DIR, "extra-files"));
+    version (Windows)
+    {
+        Vars.set("LIBEXT", ".lib");
+    }
+    else
+    {
+        Vars.set("LIBEXT", ".a");
+    }
+}
+
+string fatal(T...)(T args)
+{
+    writeln("Error: ", args);
+    return null;
+}
+string requireEnv(string name)
+{
+    const result = environment.get(name, null);
+    if (result is null)
+        assert(0, fatal("missing required environment variable '", name, "'"));
+    return result;
+}
+
+void rm(T...)(T args)
+{
+    foreach (arg; args)
+    {
+        auto expanded = expand(arg);
+        if (exists(expanded))
+        {
+            writeln("rm '", expanded, "'");
+            // Use loop to workaround issue in windows with removing
+            // executables after running then
+            for (int attempt = 1;; attempt++)
+            {
+                try {
+                    std.file.remove(expanded);
+                    break;
+                } catch (Exception e) {
+                    if (attempt >= 4)
+                        throw e;
+                    Thread.sleep(dur!"seconds"(1));
```

Comment:
how about starting with 10ms and doing an exponential back-off?

Reply:
done...except, I just did a "multiply by 2 each time" backup